api/apps/api/src/migrations/api/1685458802957-AddFeatureTagTable.ts

```diff
@@ -0,0 +1,54 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddFeatureTagTable1685458802957 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`CREATE TABLE project_feature_tags (
+      id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
+      project_id uuid NOT NULL references projects(id) ON DELETE CASCADE,
+      feature_id uuid NOT NULL references features(id) ON DELETE CASCADE,
+      tag varchar NOT NULL,
+      last_modified_at timestamp NOT NULL default now()
+    );`);
+
+    await queryRunner.query(`
+      CREATE UNIQUE INDEX single_tag_on_project_features ON project_feature_tags (project_id, feature_id);
+    `);
+
+    // A trigger that will update the tag of the incoming insert to use a capitalization equivalent tag that already exists
+    // for the same project
+    await queryRunner.query(`
+    CREATE OR REPLACE FUNCTION tr_equivalent_capitalization_tag()
+    RETURNS TRIGGER AS $BODY$
+    DECLARE
+      equivalent_tag project_feature_tags.tag%type ;
+      equivalent_tag_counter integer;
+    BEGIN
+      select count(*) into equivalent_tag_counter from project_feature_tags pft where LOWER(pft.tag) = LOWER(NEW.tag);
+      IF equivalent_tag_counter >= 1 THEN
+        select tag into equivalent_tag from project_feature_tags pft where LOWER(pft.tag) = LOWER(NEW.tag) limit 1;
+        NEW.tag = equivalent_tag;
+      ELSE
+      END IF;
+
+      RETURN NEW;
+    end
+    $BODY$ LANGUAGE plpgsql;
+
+    DROP TRIGGER IF EXISTS tr_project_feature_tags_insert ON project_feature_tags;
+
+    CREATE TRIGGER tr_project_feature_tags_insert
+    BEFORE INSERT ON project_feature_tags
```

Comment:
without `OR UPDATE` (`BEFORE INSERT OR UPDATE`), this will help when updating _the tag itself_; for example, if a user wants tagged some features in a given project with `test` and now wants to edit that tag to `Test`: in this case, a plain `update project_feature_tags set tag = 'Test' where tag = 'test' and project_id = '<pid>'` will work.

However, if a user updates a _single feature_, already tagged as `test`, now as `Test`, a capitalization inconsistency will creep in:

```
marxan-api=# select * from project_feature_tags ;
                  id                  |              project_id              |              feature_id              | tag  |      last_modified_at
--------------------------------------+--------------------------------------+--------------------------------------+------+----------------------------
 25d0526d-f119-4a97-a96f-a0b20076cbfc | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 33305ae9-340a-42d0-8b44-c05299459def | test | 2023-06-02 09:21:03.113087
 b6961ad8-aff8-41ff-ab98-b9943b28000e | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 7fd48fe5-b34e-4f1a-a21a-4c66e5562d7b | test | 2023-06-02 09:22:39.738611
 ac8ff0d9-be61-4ecb-abc9-fc4c8fde51c7 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 8fc93d5c-8a19-42a0-88aa-a519f655b0f6 | test | 2023-06-02 09:23:50.590508
 2bcca2a9-95ab-4d71-81c4-e5ce9b191862 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 1a57290d-fca6-4b00-a9f2-7776476b603b | test | 2023-06-02 09:35:48.09709
(4 rows)

marxan-api=# update project_feature_tags set tag = 'Test' where id = '2bcca2a9-95ab-4d71-81c4-e5ce9b191862';
UPDATE 1
marxan-api=# select * from project_feature_tags ;
                  id                  |              project_id              |              feature_id              | tag  |      last_modified_at
--------------------------------------+--------------------------------------+--------------------------------------+------+----------------------------
 25d0526d-f119-4a97-a96f-a0b20076cbfc | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 33305ae9-340a-42d0-8b44-c05299459def | test | 2023-06-02 09:21:03.113087
 b6961ad8-aff8-41ff-ab98-b9943b28000e | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 7fd48fe5-b34e-4f1a-a21a-4c66e5562d7b | test | 2023-06-02 09:22:39.738611
 ac8ff0d9-be61-4ecb-abc9-fc4c8fde51c7 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 8fc93d5c-8a19-42a0-88aa-a519f655b0f6 | test | 2023-06-02 09:23:50.590508
 2bcca2a9-95ab-4d71-81c4-e5ce9b191862 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 1a57290d-fca6-4b00-a9f2-7776476b603b | Test | 2023-06-02 09:35:48.09709
(4 rows)
```

If using the `OR UPDATE` for the trigger, however, updating a tag won't work without an extra step, because each time we (try to) update a tag, the trigger sets it back to the existing capitalization:

```
marxan-api=# CREATE TRIGGER tr_project_feature_tags_insert
marxan-api-#     BEFORE INSERT OR UPDATE ON project_feature_tags
marxan-api-#     FOR EACH ROW EXECUTE
marxan-api-#     PROCEDURE tr_equivalent_capitalization_tag();
CREATE TRIGGER
marxan-api=# update project_feature_tags set tag = 'test' where tag = 'Test';
UPDATE 4
marxan-api=# select * from project_feature_tags ;
                  id                  |              project_id              |              feature_id              | tag  |      last_modified_at
--------------------------------------+--------------------------------------+--------------------------------------+------+----------------------------
 2bcca2a9-95ab-4d71-81c4-e5ce9b191862 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 1a57290d-fca6-4b00-a9f2-7776476b603b | Test | 2023-06-02 09:35:48.09709
 25d0526d-f119-4a97-a96f-a0b20076cbfc | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 33305ae9-340a-42d0-8b44-c05299459def | Test | 2023-06-02 09:21:03.113087
 b6961ad8-aff8-41ff-ab98-b9943b28000e | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 7fd48fe5-b34e-4f1a-a21a-4c66e5562d7b | Test | 2023-06-02 09:22:39.738611
 ac8ff0d9-be61-4ecb-abc9-fc4c8fde51c7 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 8fc93d5c-8a19-42a0-88aa-a519f655b0f6 | Test | 2023-06-02 09:23:50.590508
(4 rows)
```
And there's other bits - for example, if creating two distinct triggers, one on insert and one on update, instead of a single one on insert or update, transparently transforming a new capitalization to the existing one... won't work as expected.

So... I'm tempted to give up and do this at application level 🤷🏼 - we can also keep this trigger-based implementation, but then we need to make sure we either improve it to avoid that updating the tag on a single feature changing only its capitalization leads to a no-op, or handle this case at the application level.

OR!!!

I think that I cracked the _consistent tag capitalization_ riddle!

```
marxan-api=# select * from project_feature_tags ;
                  id                  |              project_id              |              feature_id              | tag  |      last_modified_at
--------------------------------------+--------------------------------------+--------------------------------------+------+----------------------------
 25d0526d-f119-4a97-a96f-a0b20076cbfc | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 33305ae9-340a-42d0-8b44-c05299459def | test | 2023-06-02 09:21:03.113087
 b6961ad8-aff8-41ff-ab98-b9943b28000e | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 7fd48fe5-b34e-4f1a-a21a-4c66e5562d7b | test | 2023-06-02 09:22:39.738611
 ac8ff0d9-be61-4ecb-abc9-fc4c8fde51c7 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 8fc93d5c-8a19-42a0-88aa-a519f655b0f6 | test | 2023-06-02 09:23:50.590508
 2bcca2a9-95ab-4d71-81c4-e5ce9b191862 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 1a57290d-fca6-4b00-a9f2-7776476b603b | test | 2023-06-02 09:35:48.09709
(4 rows)
marxan-api=# create extension btree_gist;
CREATE EXTENSION
marxan-api=# alter table project_feature_tags add constraint consistent_tag_capitalization exclude using gist (tag with <>) where (project_id = project_id) deferrable initially deferred;
ALTER TABLE
marxan-api=# update project_feature_tags set tag = 'Test' where tag = 'test' and project_id = '54716cd1-be98-4bbd-8c92-3c8c6b92fbdf';
UPDATE 4
marxan-api=# select * from project_feature_tags ;
                  id                  |              project_id              |              feature_id              | tag  |      last_modified_at
--------------------------------------+--------------------------------------+--------------------------------------+------+----------------------------
 25d0526d-f119-4a97-a96f-a0b20076cbfc | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 33305ae9-340a-42d0-8b44-c05299459def | Test | 2023-06-02 09:21:03.113087
 b6961ad8-aff8-41ff-ab98-b9943b28000e | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 7fd48fe5-b34e-4f1a-a21a-4c66e5562d7b | Test | 2023-06-02 09:22:39.738611
 ac8ff0d9-be61-4ecb-abc9-fc4c8fde51c7 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 8fc93d5c-8a19-42a0-88aa-a519f655b0f6 | Test | 2023-06-02 09:23:50.590508
 2bcca2a9-95ab-4d71-81c4-e5ce9b191862 | 54716cd1-be98-4bbd-8c92-3c8c6b92fbdf | 1a57290d-fca6-4b00-a9f2-7776476b603b | Test | 2023-06-02 09:35:48.09709
(4 rows)
```

This won't help with helpfully transforming a different capitalization of a tag to its existing capitalization, but will finally enforce that an attempt to tag a feature with a different capitalization of an existing tag will fail, within any given project.

```
marxan-api=# update project_feature_tags set tag = 'test' where id = '2bcca2a9-95ab-4d71-81c4-e5ce9b191862';
ERROR:  conflicting key value violates exclusion constraint "consistent_tag_capitalization"
DETAIL:  Key (tag)=(test) conflicts with existing key (tag)=(Test).
marxan-api=# insert into project_feature_tags (project_id , feature_id, tag) values ('54716cd1-be98-4bbd-8c92-3c8c6b92fbdf', '9c1010e2-4af3-49e1-9f89-12e59829b0d3', 'test');
ERROR:  conflicting key value violates exclusion constraint "consistent_tag_capitalization"
DETAIL:  Key (tag)=(test) conflicts with existing key (tag)=(Test).
marxan-api=# insert into project_feature_tags (project_id , feature_id, tag) values ('54716cd1-be98-4bbd-8c92-3c8c6b92fbdf', '9c1010e2-4af3-49e1-9f89-12e59829b0d3', 'Test');
INSERT 0 1
```

So, I'm not sure. I'd be tempted to stop here, either doing everything at the application level (unless you have a quick idea to make the triggers work as intended on update) or use the draconian constraint (whose performance implications are not clear to me though - I _think_ that with appropriate index these `EXCLUDE` checks will have negligible overhead, but 🤔) and handle failures.

Reply:
ok, that was a lie (about the unique constraint).

I _think_ and then I'll _really_ give up that the correct one would require something like this:

```
CREATE TABLE project_feature_tags (
  id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
  project_id uuid NOT NULL references projects(id) ON DELETE CASCADE,
  feature_id uuid NOT NULL references features(id) ON DELETE CASCADE,
  tag varchar NOT NULL,
  last_modified_at timestamp NOT NULL default now(),
  tag_lower varchar generated always as (lower(tag)) stored
);

alter table project_feature_tags add constraint consistent_tag_capitalization exclude using gist (tag with <>, tag_lower with =, project_id with =) deferrable initially deferred;
```

The suggestion above would not correctly respect `project_id` boundaries.